fw/rbcx-coprocessor/src/Esp32Manager.cpp

```diff
@@ -0,0 +1,176 @@
+#include "stm32f1xx_hal.h"
+#include "stm32f1xx_hal_gpio.h"
+
+#include "FreeRTOS.h"
+#include "task.h"
+
+#include "Bsp.hpp"
+#include "utils/Debug.hpp"
+
+#include "Esp32Manager.hpp"
+
+Esp32Manager sEsp32Manager;
+
+template <void (*FreeUp)(), void (*Restore)()> class PinGuard {
+public:
+    PinGuard() { (*FreeUp)(); }
+    ~PinGuard() { (*Restore)(); }
+};
+
+#define PIN_GUARD(PIN_NUM)                                                     \
+    PinGuard<&esp32Pin##PIN_NUM##FreeUp, &esp32Pin##PIN_NUM##Restore>          \
+        _pin##PIN_NUM##guard;
+
+void __attribute__((weak)) esp32Pin0FreeUp() {}
+void __attribute__((weak)) esp32Pin0Restore() {
+    pinInit(esp0Pin, GPIO_MODE_INPUT, GPIO_NOPULL, GPIO_SPEED_FREQ_LOW);
+}
+void __attribute__((weak)) esp32Pin2FreeUp() {}
+void __attribute__((weak)) esp32Pin2Restore() {
+    pinInit(esp2Pin, GPIO_MODE_INPUT, GPIO_NOPULL, GPIO_SPEED_FREQ_LOW);
+}
+void __attribute__((weak)) esp32Pin12FreeUp() {}
+void __attribute__((weak)) esp32Pin12Restore() {
+    pinInit(esp12Pin, GPIO_MODE_INPUT, GPIO_NOPULL, GPIO_SPEED_FREQ_LOW);
+}
+void __attribute__((weak)) esp32Pin15FreeUp() {}
+void __attribute__((weak)) esp32Pin15Restore() {
+    pinInit(esp15Pin, GPIO_MODE_INPUT, GPIO_NOPULL, GPIO_SPEED_FREQ_LOW);
+}
+
+static void esp32PinENFreeUp() {
+    pinInit(espEnPin, GPIO_MODE_OUTPUT_OD, GPIO_NOPULL, GPIO_SPEED_FREQ_LOW);
+}
+
+static void esp32PinENRestore() {
+    pinInit(espEnPin, GPIO_MODE_IT_RISING, GPIO_PULLUP, GPIO_SPEED_FREQ_LOW);
+}
```

Comment:
@cednik prosím zkontroluj, jestli tady vrtím s piny dobře, hlavně ten EN pin.

Reply:
Moc mi nedává smysl to pojmenování - co má dělat FreeUp a co Restore?

Reply:
FreeUp uvolní piny, aby je bylo možné použít pro esp32 (momentálně žádný z nich nic nedělá, až na EN), Restore je pokyn, že daná periferie má pin "obnovit" do původního stavu (zavolat vlastní pinInit). To pojmenování se mi taky nelíbí, jsem otevřen nápadům :)

Reply:
Asi mám dnes nějakou nechápavou náladu :-D
Co znamená "použití pro esp32" a jaké periferie máš na mysli?

Podle mě jsou tři stavy/akce:
* ESP je v resetu, čekám na jeho uvolnění
1) chci ESP nastartovat do bootloadru - musím piny držet v jednom stavu
2) chci ESP nastartovat do uživatelského programu - musím piny držet v druhém stavu, alternativně nechat viset v luftu (možná lepší možnost, plus kontrolovat, jestli jsou v tom správném stavu a když ne, zablikat ledkama warning).
* reset uvolněn, ESP bootoje kam má
3) nad piny přebírají vládu periferie STM - prakticky jde jen o CONTROL UART, v budoucnu by to mohlo být ještě SPI pro debugování (nebo SPI pro inteligentní ledky, ale to jen v situaci, kdy v desce nebude ESP a pak je celý předchozí text irelevantní).

Reply:
"použití pro esp32" je hýbání s nimi na boostrap při resetu ESP.
"periferie" je zatím jen ControlUart

1. Normální, stav, ESP piny 0 a 2 používá ControlUart
2. Zmáčneš RESET tlačítko
3. Esp32Manager zavolá esp32PinXFreeUp (momentálně nic nemusíme dělat, tak jsou prázdné)
4. Esp32Manager použije piny na bootstrap, resetuje ESP32
5. Esp32Manager zavolá esp32PinXRestore - Control Uart si na pinech nastaví zase to, co potřebuje on

Reply:
1. Normální stav OK.
2. Tlačítko myslíš na RBCX/přímo na ESP (jsou paralelně)?
3. Se zavoláním FreeUp souhlasím - mělo by piny přenout do HI-Z (INPUT, NO_PULLUP).
4. Pustím tlačítko.
5. Piny zůstanou 1 ms v HI-Z [[ESP datasheet](https://www.espressif.com/sites/default/files/documentation/esp32_datasheet_en.pdf#5f)].
( Zde bych mohl kontrolovat stav pinů a skrz ledky hlásit, pokud piny něco drží špatně a ESP nabootoje jinam, než do uživatelské aplikace - jen pro účely debugování na první pohled).
6. Esp32Manager zavolá esp32PinXRestore a Control Uart přebere kontrolu nad piny.

Jsem li k STM připojen skrz USB, měl bych si hlídat stav RTS a DTR, jaký chce aplikace na počítači. V okamžiku, kdy DTR = 1 a RTS = 0, musí STM nastavit ESP_EN na LOW - viz tabulka vlevo dole z [[esp32_devkitc_v4-sch.pdf](https://dl.espressif.com/dl/schematics/esp32_devkitc_v4-sch.pdf)]. Z pohledu předchozích bodů se až po bod 3 včetně nic nemění a klidně to může řešit ten samý kód z přerušení - STM by mělo být úplně jedno, jestli reset ESP přišel zvenku, nebo zevnitř.
Když se tento stav RTS a DTR změní, měl bych piny opět na 1 ms nastavit do stavu podle tabulky. Dále pokračuji normálně bodem 6.
Jako nejjednodušší implementaci navrhuji definovat stav RTS a DTR když port na počítači není otevřený jako 1 1. Když připojený je, řídí se požadavky z počítače. V bodě 3 pak piny budou OPEN_DRAIN a v LOG.1 - tedy defakto HI-Z. V bodě 5 se pak stav pinů bude řídit DTR a RTS (díky OPEN_DRAIN módu pokud nastavíš logickou 0, bude tam, pokud necháš 1, STM se do stavu pinu nebude plést). Po 1 ms si piny přebere Control Uart.
Toto chování se nejvíce blíží chování "FTDI" na devkitu.

Reply:
2. je to jedno, prostě se od někud dozvím, že mám resetovat ESP (RST tlačítko, DTR/RST, watchdog...)

Možná si vůbec nerozumíme - nemám to implementované tak, že bych přímo hýbal piny kdykoliv se změní DTR/RST, to je zbytečné, já na nich pouze detekuju sekvenci, kterou s nimi zahýbá esptool při flashování - dokud žádná validní nepřijde, Esp32Manager nic nedělá.

Rozhodnutí, že Esp32Manager má resetovat ESP, ať už do bootloaderu nebo normálně tedy může přijít z těchto zdrojů:
* RESET tlačítko, ať už na ESP32 devkitu nebo na desce
* DTR/RST sekvence z esptool při flashování
* (v budoucnu Watchdog)

Jakmile některá z těchto podmínek nastane, musí:
1. to, co právě pin používá ustoupit (esp32PinXFreeUp)
2. Esp32Manager [piny přenastaví a použije na reset](https://github.com/RoboticsBrno/RB3204-RBCX/pull/31/files#diff-1a773f27fc5e076b4eeb5d88922e47c2R107)
3. pak je zase vrátit (esp32PinXRestore)

Reply:
HIGH na ESP pinu 12 během jeho resetu je chyba - opravím.
Strapping piny nechci mít jako PP - vem si situaci, kdy při resetu ESP na něm budu zároveň držet BOOT tlačítko - tím chytím IO0 k zemi. OD je lepší.

V implementaci, kterou jsi poslal, esp32PinXFreeUp s pinem opravdu nemusí udělat nic, protože [Esp32Manager::executeReset](https://github.com/RoboticsBrno/RB3204-RBCX/pull/31/files#diff-1a773f27fc5e076b4eeb5d88922e47c2R88) si ty piny prostě zabaví a na nic se neptá.
Přijde mi ale, že mít takovouhle prázdnou funkci je zavádějící - člověk se pak musí začít nořit do kódu, aby zjistil, kde se ty piny valstně uvolňují, kdy chce zjistit, jak to funguje (a proč to třeba nedělá to, co by od toho čekal).
Smysluplné využití té funkce mě napadá jedině flushnout buffery Control Uartu, aby někde nezůstalo něco vyset. Ale pak myslím stačí jedna Free a jedna Restore, nemusí ji mít každý pin samostatně.

K detekci resetovací sekvence: využíval jsem toho, že v Lorris z terminálu můžu zatržítkem DTR resetovat ESP. Detekovat konkrétní sekvenci esptoolu mi přijde zbytečná práce navíc a připraví nás to o tuhle featurku.

Reply:
Ok, hodím je na open drain, to dává smysl :)

Je to tak, přesně na ty buffery je myšlená. Je jich více, protože s Esp32Manageru nevím, kdo který pin používá. IO0 může být UART, 2 ultrazvuky, 12 serva, 15 motory...

To stále můžeš, esptool [dělá to stejné](https://github.com/espressif/esptool/blob/master/esptool.py#L1044). Reálně ty DTR/RST stejně přímo na ESP32 piny nejdou ani na devkitu - jinak bys mohl držet ESP32 v resetu když necháš RTS na high, což se neděje.

Reply:
Dobře, nechme teda uklízecí funkce pro každý pin zvlášť, asi to dává smysl.

Mít definované defaultní prázdné weak funkce se mi nelíbí - přijde mi ta vazba "skrytá". Nějaké callback funkce, které si registruješ, by mi přišly přehlednější. Ale z pohledu optimalizace rychlosti běhu kódu to dává smysl.

Inicializace pinů u konkrétní periferie je prakticky jen u Uartů, zbytek to má v pinInit v Bsp (aspoň v poslední verzi, na kterou jsem koukal a to už je dlouho).
Z pohledu softwaru je hezké mít inicializaci pinů společně s danou periferií. Z pohledu hardwaru to dává smysl mít hned na začátku všechno najednou.
Asi se kloním k variantě mít to u periferie. Konkrétně u Uartu bych piny předával v konstruktoru a pak měl metody open (bez parametrů zachová poslední konfiguraci) a close, přišemž close Uart od pinů odpojí (vrátí řízení GPIO) a open připojí. V těch uklízecích funkcích pak volat jen close a open.

Ano, pomocí RST ho můžu reetovat, ale v současné implementaci tím vygeneruji na EN jen puls, nemůžu ho v resetu držet, zož se občas hodí :-)

K původní otázce: piny 0 a 2 bys teda v Restore měl připojit zpět na periferii, ne je nastavit na input, ne?
Pro EN pin podlě mě Free a Restore nechceš - konfigurace toho pinu se neměmí, je stále OD, jen ho chceš přerušením monitorovat a při požadavku na reset na chvíli chytit do nuly.

Reply:
> Mít definované defaultní prázdné weak funkc...

Ty weak funkce jsou takhle, protože je to během compile time, zadarmo. Taky bych byl radši za lepší řešení. Pokud oželíme compile time, co toto míst všech současných `pinInit` volání, např. v ControlUart

```cpp
sPinRegistry.assign(<GPIOD, PIN2>, []() {
    return PinConfig(GPIO_MODE_OUTPUT_OD, GPIO_NOPULL);
});
```

Ke všem pinům tedy máme v jedom místě registrované jejich "Restore" funkce, kdybych měl použít názvy z tohoto PR. EspManager si od tam může danou restore funkci najít podle čísla pinu na STM (narozdíl od stavu v tomto PR, kde periferie musí vědět, že nějaký STM pin je současně bootstrap pin na ESP32). FreeUp funkci by si také mohla periferie registrovat, pokud ji potřebuje.

Pokud s tímto souhlasíš, tak bych to ale nechal na další PR, je to velká změna.

> Inicializace pinů u konkrétní periferie je prakticky jen u Uartů...

Momentálně uarty nejsou vůbec generické, každá je svá vlastní věc (ony jsou taky všechny nastavené jinak ohledně DMA a interruptů). Začali jsme už s tímhle spartánským designem, teď rozhodně není čas to měnit :shrug: 

> Ano, pomocí RST ho můžu reetovat, ale v současné implementaci tím vygeneruji na EN jen puls, nemůžu ho v resetu držet, zož se občas hodí :-)

Je to tak. Ještě se v tom zkusím povrtat.

> K původní otázce: piny 0 a 2 bys teda v Restore měl připojit zpět na periferii, ne je nastavit na input, ne?

Ty Restore, co nastavují na INPUT, jsou weak, a v ControlLink jsou pro piny 0 a 2 přepsané. Ty defaultní weak implementace nastavují piny do výchozího stavu (jako při startu STM) protože jsem potřeboval otestovat, jestli mi to resetování funguje správně (kdyby byly prázdné, zůstali by piny 12 a 15 ve stavu který jsem jim nastavil pro reset esp).

> Pro EN pin podlě mě Free a Restore nechceš ..

Ano, ty pro EN pin jsou static, tzn pouze pro ten jeden soubor - existují, abych na EN pin mohl použít PIN_GUARD makro stejně jako na ty ostatní.

Reply:
take/release

Ale vazba přes weak funkce mi taky moc nevoní. Co je mít v Bsp a tam vázat na konkrétní věc?

```cpp
void esp32Pin15TakeRelease(bool take) {
  if (take) {
    věcClose();
  } else {
    věcOpen();
  }
}
```

Reply:
Můžem to teď nechat

Reply:
> Případně to stejné, ale ve formě jednoho pole v Bsp.hpp, indexovaného podle čísla portu + pinu, taková tabulka. To bychom pak nepotřebovali dynamickou alokaci paměti a bylo by to taky compile time.

V té tabulce by bylo defaultní nastavení pinu, podle kterého by se dělala prvnotní inicializace a EspManager by změněné piny podle ní reinicializoval? To se mi docela líbí.

> Pokud s tímto souhlasíš, tak bych to ale nechal na další PR, je to velká změna.

Jj, jsem pro.
 
> Momentálně uarty nejsou vůbec generické, každý je svá vlastní věc (ony jsou taky všechny nastavené jinak ohledně DMA a interruptů). Začali jsme už s tímhle spartánským designem, teď rozhodně není čas to měnit 🤷

Ok, teď to neřešme.
 
> Je to tak. Ještě se v tom zkusím povrtat.

👍

> Ty Restore, co nastavují na INPUT, jsou weak, a v ControlLink jsou pro piny 0 a 2 přepsané. Ty defaultní weak implementace nastavují piny do výchozího stavu (jako při startu STM) protože jsem potřeboval otestovat, jestli mi to resetování funguje správně (kdyby byly prázdné, zůstali by piny 12 a 15 ve stavu který jsem jim nastavil pro reset esp).

Tohle vyřeší tabulka s defaultním nastavením.

> Ano, ty pro EN pin jsou static, tzn pouze pro ten jeden .cpp soubor - existují, abych na EN pin mohl použít PIN_GUARD makro stejně jako na ty ostatní.

Ok, tak důkladně jsem to nečetl :-)

Reply:
Prohlašuju za vyřešené :)